src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1202,6 +1209,95 @@ DatabaseShardImp::findAcquireIndex(
     return boost::none;
 }
 
+void
+DatabaseShardImp::scheduleFinalizeShard(std::uint32_t shardIndex)
+{
+    taskQueue_->addTask([this, shardIndex]
+    {
+        auto retry = [this, shardIndex]
+        {
+            const uint8_t TIMER_INTERVAL = 60;
+
+            using timer = boost::asio::basic_waitable_timer<
+                std::chrono::steady_clock>;
+
+            std::shared_ptr<timer> retryTimer(new timer(app_.getIOService()));
+
+            retryTimer->expires_from_now(std::chrono::seconds(TIMER_INTERVAL));
+            retryTimer->async_wait([this, shardIndex, retryTimer]
+                (boost::system::error_code const& ec)
+            {
+                scheduleFinalizeShard(shardIndex);
+            });
+        };
+
+        // We use the presence (or absence) of the validated
+        // ledger as a heuristic for determining whether or
+        // not we have stored a ledger that comes after the
+        // last ledger in this shard. A later ledger must
+        // be present in order to reliably retrieve the hash
+        // of the shard's last ledger.
+        if (app_.getLedgerMaster().getValidatedLedger())
+        {
+            auto const hash = app_.getLedgerMaster().walkHashBySeq(
+                lastLedgerSeq(shardIndex));
+
+            std::lock_guard lock(mutex_);
+            auto const it {shards_.find(shardIndex)};
+
+            if (auto found = (it == shards_.end()); found ||
+                it->second.state != ShardInfo::State::postImport)
+            {
+                JLOG(j_.error()) <<
+                    "Shard " << shardIndex << " failed to import";
+                return;
+            }
+
+            if (!hash || hash->isZero())
+            {
+                if (it->second.attemptHashConfirmation())
+                {
+                    JLOG(j_.warn())
+                        << "Retrying last ledger hash confirmation"
+                           " for shard " << shardIndex << ".";
+
+                    retry();
+                }
+                else
+                {
+                    JLOG(j_.error())
+                        << "Shard "
+                        << shardIndex
+                        << " failed to import."
+                        << " Maximum last ledger hash"
+                           " confirmation attempts reached.";
+
+                    removeFailedShard(it->second.shard, lock);
+                }
+                return;
+            }
+
+            it->second.lastLedgerHash = *hash;
+            finalizeShard(it->second, true, lock);
+        }
+
+        // Our heuristic technique didn't check out.
+        // Try again when our timer expires.
+        else
+        {
+            JLOG(j_.warn())
+                << "Failed to retrieve validated ledger while importing shard "
+                << shardIndex
+                << ". Retrying on timer expiration.";
```

Comment:
nit: to maintain consistency with source file. I'd also drop severity on this one to debug.
```
            JLOG(j_.debug()) <<
                "shard " << shardIndex <<
                " no validated ledger while importing shard," <<
                " retrying last ledger hash confirmation";
```

Reply:
Removed